### PR TITLE
feat: export api spec

### PR DESCRIPTION
**Pull request by wenytang-ms**
demo1

g1

### CODE REVIEW

File: src/constants.ts [hunk 1]
```diff
@@ -34,3 +34,7 @@ export const ApiSpecificationOptions = {
     localFile: UiStrings.ApiSpecificationOptionLocalFile,
     activeEditor: UiStrings.ApiSpecificationOptionActiveEditor,
 };
+export const DefinitionFormat = {
+    inline: 'inline',
+    link: 'link',
+}
```
Comment (formulahendry) [inline src/constants.ts:40]: Fix the lint error.

File: package.json [hunk 2]
```diff
@@ -234,7 +234,7 @@
           "when": "never"
         },
         {
-          "command": "azure-api-center.exportOpenApi",
+          "command": "azure-api-center.exportApi",
           "when": "never"
```
Comment (formulahendry) [inline package.json:238]: For newly-added commands. let's support triggering it from commandPalette

File: src/commands/exportApi.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        node?.subscription!,
+        getResourceGroupFromId(node?.id!),
+        node?.apiCenterName!);
+
+    const exportedSpec = await apiCenterService.exportSpecification(
+        node?.apiCenterApiName!,
+        node?.apiCenterApiVersionName!,
+        node?.apiCenterApiVersionDefinition.name!);
+
+    await writeToHttpFile(node!, exportedSpec.format, exportedSpec.value);
```
Comment (formulahendry) [inline src/commands/exportApi.ts:27]: writeToHttpFile?

File: package.json (hunk 2 above)
Comment (wenytang-ms) [reply package.json]: so we need to also support command palette to trigger the exportApi and let user select the node?

File: src/commands/exportApi.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import { DefinitionFormat, openapi } from "../constants";
+import { ext } from "../extensionVariables";
+import { ApiVersionDefinitionTreeItem } from "../tree/ApiVersionDefinitionTreeItem";
+import { createTemporaryFolder } from "../utils/fsUtil";
+export namespace ExportAPI {
+    export async function exportApi(
+        context: IActionContext,
+        node?: ApiVersionDefinitionTreeItem): Promise<void> {
+        if (!node) {
+            node = await ext.treeDataProvider.showTreeItemPicker<ApiVersionDefinitionTreeItem>(`${ApiVersionDefinitionTreeItem.contextValue}-${openapi}`, context);
```
Comment (formulahendry) [inline src/commands/exportApi.ts:18]: We need to be able to select any definition, not only openai.
(Note: the showTreeItemPicker support searching by Regex)

File: src/commands/exportApi.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    function getFolderName(treeItem: ApiVersionDefinitionTreeItem): string {
+        return `${treeItem.apiCenterName}-${treeItem.apiCenterApiName}`;
+    }
+
+    function getFilename(treeItem: ApiVersionDefinitionTreeItem): string {
+        return `${treeItem.apiCenterApiVersionDefinition.name}`;
+    }
+
+    async function writeToTempFile(node: ApiVersionDefinitionTreeItem, specFormat: string, specValue: string) {
+        if (specFormat === DefinitionFormat.inline) {
```
Comment (formulahendry) [inline src/commands/exportApi.ts:41]: What if it is not inline? Our extension will do nothing?

File: src/test/unit/commands/exportApi.test.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                    specification: {
+                        name: "fakeName"
+                    }
+                }
+            } as ApiCenterApiVersionDefinition
+        );
+        sandbox.stub(node, "subscription").value("fakeSub");
+        sandbox.stub(node, "id").value("/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/test/providers/Microsoft.ApiCenter/services/test/workspaces/default/apis/test/versions/v1/definitions/openapi");
+
+    })
```
Comment (formulahendry) [inline src/test/unit/commands/exportApi.test.ts:54]: fix lint warning.

File: src/commands/exportApi.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            await ExportAPI.showTmpFile(node, specValue);
+        }
+    }
+
+    export async function showTmpFile(node: ApiVersionDefinitionTreeItem, fileContent: string) {
```
Comment (formulahendry) [inline src/commands/exportApi.ts:46]: Minor: better to use 'temp' or 'tmp' in all places.

File: src/constants.ts [hunk 8]
```diff
@@ -34,3 +34,7 @@ export const ApiSpecificationOptions = {
     localFile: UiStrings.ApiSpecificationOptionLocalFile,
     activeEditor: UiStrings.ApiSpecificationOptionActiveEditor,
 };
+export const DefinitionFormat = {
```
Comment (formulahendry) [inline src/constants.ts:37]: Better to put into https://github.com/microsoft/vscode-azureapicenter/blob/dev/src/azure/ApiCenter/contracts.ts

File: src/commands/exportApi.ts (hunk 5 above)
Comment (wenytang-ms) [reply src/commands/exportApi.ts:41]: currently, there not have any link api definitions in server side, also post question in API Center CLI team channel, has no response.

File: src/commands/exportApi.ts (hunk 4 above)
Comment (wenytang-ms) [reply src/commands/exportApi.ts]: thanks! updated.

File: src/commands/exportApi.ts (hunk 7 above)
Comment (wenytang-ms) [reply src/commands/exportApi.ts]: all turn to temp

File: src/constants.ts (hunk 8 above)
Comment (wenytang-ms) [reply src/constants.ts]: updated

File: src/commands/exportApi.ts [hunk 9]
```diff
@@ -15,7 +15,7 @@ export namespace ExportAPI {
         context: IActionContext,
         node?: ApiVersionDefinitionTreeItem): Promise<void> {
         if (!node) {
-            node = await ext.treeDataProvider.showTreeItemPicker<ApiVersionDefinitionTreeItem>(`${ApiVersionDefinitionTreeItem.contextValue}-${openapi}`, context);
+            node = await ext.treeDataProvider.showTreeItemPicker<ApiVersionDefinitionTreeItem>(new RegExp(`${ApiVersionDefinitionTreeItem.contextValue}*`), context);
```
Comment (formulahendry) [inline src/commands/exportApi.ts:19]: Will `*` work? I think it should be `.*`.

**Review by formulahendry — CHANGES_REQUESTED** [review]
Fix the failed UT.
BTW, in future, we'd better wrap those general mock (e.g. mock TelemetryClient) into a single function so that they could be shared in each test case.

**Comment by wenytang-ms** [comment]
> Fix the failed UT. BTW, in future, we'd better wrap those general mock (e.g. mock TelemetryClient) into a single function so that they could be shared in each test case.

updated, just mock the telemetryclient event

File: src/commands/exportApi.ts (hunk 9 above)
Comment (wenytang-ms) [reply src/commands/exportApi.ts:19]: yes, `*` works as `.*`

**Review by formulahendry — APPROVED** [review]

**Pull request merged**